### ISSUE TITLE
Spurious covariance error in curried function decorator checking

### ISSUE THREAD

**Issue by kwshi** [microsoft/pyright#4274]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
I can't explain what the general issue is, but here's the example. I'm trying to define a decorator that post-processes a function's output, defined like so:
```python
A = typing.TypeVar("A")
B = typing.TypeVar("B")
P = typing.ParamSpec("P")

def transform(
    transformer: typing.Callable[[A], B]
) -> typing.Callable[[typing.Callable[P, A]], typing.Callable[P, B]]:
    def wrap(f: typing.Callable[P, A]) -> typing.Callable[P, B]:
        def wrapped(*args: P.args, **kwargs: P.kwargs):
            return transformer(f(*args, **kwargs))

        return wrapped

    return wrap
```

Then one would use it like so:
```python
def add1(n: int) -> int: return n+1

@transform(add1)
def square(n: int) -> int: return n*n

square(3) # returns 3*3 + 1 -> 10
```

This example actually works great. The issue arises in a more complex setting where I'm trying to post-process an iterable with `sum`. Basically, the following example doesn't work:
```python
@transform(sum)
def ints() -> list[int]:
  return [1, 2, 3]

# the above is same as
def ints_list() -> list[int]:
  return [1, 2, 3]
ints = transform(sum)(ints_list)
```

Weirdly enough, the error is resolved if I define a custom version of `sum` that is monomorphic (only takes `list[int]`); the error is also resolved if instead of defining `transform` as a curried function I define it uncurried (to be used as `transform(sum, ints_list)`).

**To Reproduce**
See full program below.

**Expected behavior**

See program below.

Ideally all four examples should type-check correctly. Though it might be possible I'm not giving `transform` the correct type annotations (though I don't think that's the case?); in that case there should at least be consistency across the four examples--they should all succeed or all fail. The fact that one of them errors while all other three succeed seems like a bug to me.

**Screenshots or Code**

```python
import typing

A = typing.TypeVar("A")
B = typing.TypeVar("B")
P = typing.ParamSpec("P")


# a decorator that post-processes the decorated function with `transformer`.
# type signature in ML notation: (a -> b) -> ((params -> a) -> (params -> b)).
def transform(
    transformer: typing.Callable[[A], B]
) -> typing.Callable[[typing.Callable[P, A]], typing.Callable[P, B]]:
    def wrap(f: typing.Callable[P, A]) -> typing.Callable[P, B]:
        def wrapped(*args: P.args, **kwargs: P.kwargs):
            return transformer(f(*args, **kwargs))

        return wrapped

    return wrap


# same thing, just not in the form of a decorator.
def transform_uncurried(
    transformer: typing.Callable[[A], B], f: typing.Callable[P, A]
) -> typing.Callable[P, B]:
    def wrapped(*args: P.args, **kwargs: P.kwargs):
        return transformer(f(*args, **kwargs))

    return wrapped


def my_sum(i: typing.Iterable[int]):
    return sum(i)


def simple_iterator() -> list[int]:
    return [1, 2, 3]


# in practice I would use this via `@transform(sum)`, etc., but directly
# applying it here for demonstration purposes (to save space).
x = transform(sum)(simple_iterator)  # ERROR
y = transform(my_sum)(simple_iterator)  # OK

z = transform_uncurried(sum, simple_iterator)  # OK
w = transform_uncurried(my_sum, simple_iterator)  # OK
```

**VS Code extension or command-line**
Using `pyright` 1.1.280 + Neovim LSP extension

**Comment by erictraut** [comment]
The problem here is that `sum` is an overloaded function. When it is assigned to the argument corresponding to the `transformer` parameter (which has a declared type of `Callable[[A], B]`), only the first compatible overload of `sum` is captured by the Callable. There is no provision in the type system to capture an overload, so type checkers capture the first overload that is a match for the type.

As you pointed out, you can work around this by declaring your own monomorphic version of `sum`.

**Comment by kwshi** [comment]
Okay, but why does the uncurried version not have this issue? Is it because when it's uncurried, the second argument may be used to assist in inferring the correct overload of the first argument?

**Comment by erictraut** [comment]
In the uncurried version, there is additional context that allows the type checker to pick a different overload of `sum`. The additional context comes from the fact that type variable `A` appears in both the `transformer` and `f` parameter annotations. The constraint solver finds an overload that satisfies `A` in both cases. In the uncurried example, there is no additional constraint, and multiple overloads match the type `Callable[[A], B]`. The first matching overload is chosen.

**Comment by kwshi** [comment]
I understand; thanks for the explanation!
